### DISCUSSION TITLE
Add Support for Fibers (via JS Generators)

### DISCUSSION THREAD

**Discussion by AnthonySuper** [opal/opal#2207] [General]
[JavaScript generators](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Iterators_and_Generators) have [increasingly good support](http://kangax.github.io/compat-table/es5/) among browsers. They're not 100% supported yet, but they will hopefully be soon.

I believe that it should be possible to implement much of Ruby's `Fiber` module using Generators. The semantics aren't exactly the same, but both allow for the same kind of suspended execution. Not only would this allow us to make the implementation of `Enumerable` slightly cleaner, but this could pave the way to an eventual implementation of an [async/await-like](https://ponyfoo.com/articles/understanding-javascript-async-await) library using Fibers, which would make request-based code much easier.

I don't really know exactly where we'd start with doing this, but I'd be willing to help in any way I can.

**Comment by meh** [comment]
This sounds fun, I'll give it a shot.

**Comment by AnthonySuper** [comment]
If you need any assistance, let me know.

I have done some thinking about this, so have a little brain dump about possible pain points 
1. Ruby uses `Fiber.yield` to transfer control, JS uses the `yield` keyword. This may mean that we'll have to special-case `Fiber.yield` in the parser, which is less-than-ideal, as I could see somebody doing something weird like calling `yield` on a variable which may sometimes be the constant `Fiber`
2. JS generators don't have an equivalent of `Fiber.transfer` that I can see.
3. Exceptions should work relatively transparently, since Ruby and JS seem to handle them in the same way (bubbling them up to the calling context of the fiber/generator)
4. The way arguments may be passed to fibers seems fairly transparent as well. In ruby you use `fiber.resume(arg)`, and in JS you use `fiber.next(arg)`
5. It may be possible to implement `await` within the `Promise` library via Fibers. If this turns out to be true, it would make using Opal a lot easier, since you could pretend that code is synchronous. Using the library's example, this:

``` ruby
HTTP.get("user_details").then do |response|
  HTTP.get("user_contacts?id=#{response.json[:user_id]}")
end.then do |response|
  HTTP.get("user_details?id=#{response.json[:contact_id]}")
end.then do
  puts "All finished"
end
```

May be translatable into something like this:

``` ruby
response = HTTP.get("user_details").await
contacts = HTTP.get("user_contacts?id=#{response.json[:user_id]}").await
details = HTTP.get("user_details?id=#{response.json[:contact_id]}").await
puts "All finished"
```

Which seems much cleaner.
6. Once Fibers are working, I think it would be beneficial to implement `Enumerable` using them, which is how I believe the Ruby standard library does it. This is a patch I'm fairly certain I actually _would_ be able to contribute.

**Comment by meh** [comment]
Ok first pass of experimentation and reading.

Issues:
- The `yield` keyword must be present within the body of the generator function, this means calling `Fiber.yield` within another block won't work, at all.
- This means `Fiber.new` and `Fiber.yield` cannot be implemented like other things and will require compiler support to convert `Fiber.new` into a `function*` declaration with any call to `Fiber.yield` compiled as a barebone `yield`.
- This means I'm not entirely sure how useful this would be in practice since looping is done with block calls.

Marking this for discussion for further ideas, implementing this in the compiler _shouldn't_ be too hard, it's just a matter of special casing `Fiber.new` and `Fiber.yield` and avoid creating intermediary functions or restricting the semantics.

Main issue is how useful this would actually be since you wouldn't be able to call `Fiber.yield` within another block.

**Comment by meh** [comment]
@AnthonySuper damn, you beat me by few seconds :panda_face: 

The main issue I see is the fact you can't call yield at any other call depth than the root of the generator function, any ideas to work around that?

**Comment by AnthonySuper** [comment]
Well, after screwing around with it for a while, I don't have any easy answers, unfortunately. A common use case of fibers is something like:

``` ruby
fiber = Fiber.new do
  some_array.each do |elem|
    Fiber.yield elem
  end
end
```

You can sort of fake this by doing something like this:

``` js
var Fiber = {
  yieldValues: [],
  yield: function(val) {
    if(val) {
      this.yieldValues.push(val);
    }
  },
  [Symbol.iterator]: function*() {
    yield* this.yieldValues[Symbol.iterator]();
  }
}

some_array = [1, 2, 3];

function *gen(){
  some_array.forEach(function(elem) {
    Fiber.yield(elem);
  });
  yield* Fiber;
}

var f = gen();
var q = f.next();
while(! q.finished) {
  console.log(q);
  q = f.next();
}
```

However, this ruins the entire point of Fibers, which is user-controlled scheduling, so it's pretty much useless. Implementing this might not actually be possible, which seems pretty unfortunate.

**Comment by meh** [comment]
Yeah, it seems to be a dead end for now, I'll leave the issue open in case someone comes up with a good idea.

**Comment by iliabylich** [comment]
Almost implemented it locally but then realized that it's not that easy :smile: 

I've started with [the following code](http://opalrb.org/try/?code:closure_value%20%3D%202%0A%0Aif%20RUBY_ENGINE%20%3D%3D%20%27opal%27%0A%20%20class%20Fiber%0A%20%20%20%20def%20initialize%28%26block%29%0A%20%20%20%20%20%20%60self.%24%24block%20%3D%20block%60%0A%20%20%20%20end%0A%0A%20%20%20%20def%20resume%28value%20%3D%20nil%29%0A%20%20%20%20%20%20%25x%7B%0A%20%20%20%20%20%20%20%20if%20%28!self.hasOwnProperty%28%27%24%24generator%27%29%29%20%7B%0A%20%20%20%20%20%20%20%20%20%20self.%24%24generator%20%3D%20self.%24%24block%28value%29%3B%0A%20%20%20%20%20%20%20%20%7D%0A%0A%20%20%20%20%20%20%20%20return%20self.%24%24generator.next%28%29.value%3B%0A%20%20%20%20%20%20%7D%0A%20%20%20%20end%0A%20%20end%0A%0A%0A%20%20block%20%3D%20%25x%7B%0A%20%20%20%20function*%20%28%29%20%7B%0A%20%20%20%20%20%20var%20i%20%3D%201%3B%0A%0A%20%20%20%20%20%20while%28true%29%20%7B%0A%20%20%20%20%20%20%20%20yield%20i%20*%20closure_value%3B%0A%20%20%20%20%20%20%20%20i%2B%2B%3B%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%7D%0A%20%20%7D%0Aelse%0A%20%20%23%20Ruby%20equivalent%3A%0A%20%20block%20%3D%20proc%20%7B%20i%20%3D%201%3B%20loop%20%7B%20Fiber.yield%20i%20*%20closure_value%3B%20i%20%2B%3D%201%20%7D%20%7D%0Aend%0A%0Af%20%3D%20Fiber.new%28%26block%29%0Ap%20f.resume%0Ap%20f.resume%0Ap%20f.resume%0Ap%20f.resume), for some reason I thought that it may cover all cases.

I've added two AST rewriters that convert:
- `Fiber.new {}` to a separate node `:fiber_new` (required for `function *()`)
- block passed to `Fiber.yield` to a separate node `:fiber_iter` (required for `yield`)

So, `opal --sexp -e "Fiber.new { Fiber.yield 123 }"` became:

```
s(:send,
  s(:const, nil, :Fiber), :new,
  s(:fiber_iter,
    s(:args),
    s(:fiber_yield,
      s(:int, 123))))
```

Both nodes that compile these new nodes to JS are quite simple like this:

``` ruby
module Opal
  module Nodes
    class FiberYieldNode < Base
      handle :fiber_yield
      chlidren :value

      def compile
        push "yield ", expr(value), ";"
      end
    end
  end
end

```

I've made it working with simple cases like `Fiber.new { Fiber.yield 123 }.resume`. But then I've tried:

``` ruby
f = Fiber.new do
  i = 1
  loop { Fiber.yield(i); i += 1 }
end
```

Generated code doesn't work.
- `Fiber.new` takes a `function *`
- The body of this function invokes `Kernel#loop` which takes a `function` (a block that is passed to `loop`)
- This function invokes `yield` inside.
- This is wrong

From what I see this code can't work:

``` js
a = function *() {
  (function() { yield 123 })()
}

b = a()
b.next()
```

Any ideas how to implement `Fiber.yield` inside a block?

**Comment by AnthonySuper** [comment]
I think that Fiber.yield may not be something we can implement easily inside a block, unfortunately, because Javascript generators are heavily limited and sort of awkward. Like JS in general.

**Comment by zetachang** [comment]
The truth is that JS doesn't support coroutines,  and simulate Fiber using either `async` `await` or generator will have runtime overhead given that all the functions nested inside needs to be marked as async .

A possible approach might be compiling two versions of methods (generator and normal function) and decide which to run either at runtime or compile time.
